src/store/uploads.tsx

```diff
@@ -0,0 +1,240 @@
+  updateAsset: (contentId: string, lastStatus?: AssetUploadStatus) => void
+  removeAsset: (contentId: string) => void
+  uploadsProgress: UploadsProgressRecord
+  notifications: {
+    uploading: number
+    uploaded: number
+  }
+  assetFiles: AssetFile[]
+  navigate: NavigateFunction | null
+  startFileUpload: (
+    file: File | Blob | null,
+    asset: InputAssetUpload,
+    storageUrl: string,
+    opts?: StartFileUploadOptions | undefined
+  ) => Promise<void>
+}
+
+export const useUploadsManagerStore: StateCreator<UploadStoreState & Partial<SnackbarState>> = (set, get, api) => ({
+  uploadsState: [],
+  navigate: null,
+  notifications: {
+    uploaded: 0,
+    uploading: 0,
+  },
+  uploadsProgress: {},
+  assetFiles: [],
+  startFileUpload: async (file, asset, storageUrl, opts) => {
+    const displayUploadingNotification = debounce(() => {
+      get().displaySnackbar?.({
+        title:
+          get().notifications.uploading > 1
+            ? `${get().notifications.uploading} assets being uploaded`
+            : 'Asset being uploaded',
+        iconType: 'info',
+        timeout: UPLOADING_SNACKBAR_TIMEOUT,
+        actionText: 'See',
+        onActionClick: () => get().navigate?.(absoluteRoutes.studio.uploads()),
+      })
+      set((state) => ({ ...state, notifications: { ...state.notifications, uploading: 0 } }))
+    }, 700)
+
+    const displayUploadedNotification = debounce(() => {
+      get().displaySnackbar?.({
+        title: get().notifications.uploaded > 1 ? `${get().notifications.uploaded} assets uploaded` : 'Asset uploaded',
+        iconType: 'success',
+        timeout: UPLOADED_SNACKBAR_TIMEOUT,
+        actionText: 'See',
+        onActionClick: () => get().navigate?.(absoluteRoutes.studio.uploads()),
+      })
+      set((state) => ({ ...state, notifications: { ...state.notifications, uploaded: 0 } }))
+    }, 700)
+
+    const setAssetUploadProgress = (progress: number) => {
+      set((stat) => ({
+        ...stat,
+        uploadsProgress: { ...stat.uploadsProgress, [asset.contentId]: progress },
+      }))
+    }
+    const fileInState = get().assetFiles?.find((file) => file.contentId === asset.contentId)
+    if (!fileInState && file) {
+      set((state) => ({ ...state, assetFiles: [...state.assetFiles, { contentId: asset.contentId, blob: file }] }))
+    }
+
+    rax.attach()
+    const assetUrl = createStorageNodeUrl(asset.contentId, storageUrl)
+
+    try {
+      if (!fileInState && !file) {
+        throw Error('File was not provided nor found')
+      }
+
+      if (!opts?.isReUpload && file) {
+        get().addAsset({ ...asset, lastStatus: 'inProgress', size: file.size })
+      }
+
+      setAssetUploadProgress(0)
+
+      const setUploadProgressThrottled = throttle(
+        ({ loaded, total }: ProgressEvent) => {
+          get().updateAsset(asset.contentId, 'inProgress')
+          setAssetUploadProgress((loaded / total) * 100)
+        },
+        3000,
+        { leading: true }
+      )
```

Comment:
It seems like introducing state management resolves the issue of upload progress causing multiple rerenders, so that's a good thing. We will not need that throttle then